scipy/stats/_sensitivity_analysis.py

```diff
@@ -0,0 +1,466 @@
+    """Sample two matrices A and B.
+
+    Uses a Sobol' sequence with 2`d` columns to have 2 uncorrelated matrices.
+    This is more efficient than using 2 random draw of Sobol'.
+    See sec. 5 from [1]_.
+
+    References
+    ----------
+    .. [1] Saltelli, A., P. Annoni, I. Azzini, F. Campolongo, M. Ratto, and
+       S. Tarantola. "Variance based sensitivity analysis of model
+       output. Design and estimator for the total sensitivity index."
+       Computer Physics Communications, 181(2):259-270,
+       :doi:`10.1016/j.cpc.2009.09.018`, 2010.
+    """
+    d = len(dists)
+    A_B = qmc.Sobol(d=2*d, seed=random_state, bits=64).random(n)
+
+    A, B = A_B[:, :d], A_B[:, d:]
+
+    for d_, dist in enumerate(dists):
+        dist_rng = NumericalInversePolynomial(
+            dist, random_state=random_state
+        )
+        A[:, d_] = dist_rng.ppf(A[:, d_])
+        B[:, d_] = dist_rng.ppf(B[:, d_])
+
+    return A, B
+
+
+def sample_AB(A: np.ndarray, B: np.ndarray) -> np.ndarray:
+    """AB matrix.
+
+    AB: columns of B into A. Shape (n*d, d). e.g in 2d
+    Take A and replace 1st column with 1st column of B. You have (n, d)
+    Then A and replace 2nd column with 2nd column of B. You have (n, d)
+    Concatenate to get AB. Which means AB is (2*n, d).
+    """
+    n, d = A.shape
+    AB = np.empty((int(d*n), d))
+    for i in range(d):
+        AB[i*n:(i+1)*n, :] = np.column_stack((A[:, :i], B[:, i], A[:, i+1:]))
+
+    return AB
+
+
+def saltelli_2010(
+    f_A: np.ndarray, f_B: np.ndarray, f_AB: np.ndarray
+) -> Tuple[np.ndarray,  np.ndarray]:
+    r"""Saltelli2010 formulation.
+
+    .. math::
+
+        S_i = \frac{1}{N} \sum_{j=1}^N
+        f(\mathbf{B})_j (f(\mathbf{AB}^{(i)})_j - f(\mathbf{A})_j)
+
+    .. math::
+
+        S_{T_i} = \frac{1}{N} \sum_{j=1}^N
+        (f(\mathbf{A})_j - f(\mathbf{AB}^{(i)})_j)^2
+
+    Parameters
+    ----------
+    f_A, f_B, f_AB : array_like (n, x)
+        Function evaluations ``n`` being the number of evaluations and ``x``
+        a vector output.
+
+    Returns
+    -------
+    s, st : array_like (x, d)
+        First order and total order Sobol' indices.
+
+    References
+    ----------
+    .. [1] Saltelli, A., P. Annoni, I. Azzini, F. Campolongo, M. Ratto, and
+       S. Tarantola. "Variance based sensitivity analysis of model
+       output. Design and estimator for the total sensitivity index."
+       Computer Physics Communications, 181(2):259-270,
+       :doi:`10.1016/j.cpc.2009.09.018`, 2010.
+    """
+    f_AB = f_AB.reshape(-1, *f_A.shape)
+
+    var = np.var(np.vstack([f_A, f_B]), axis=0)
+
+    s = np.mean(f_B * (f_AB - f_A), axis=1) / var
+    st = 0.5 * np.mean((f_A - f_AB) ** 2, axis=1) / var
+
+    return s.T, st.T
+
+
+@dataclass
+class SobolResult:
+    first_order: np.ndarray
+    total_order: np.ndarray
+    _A: np.ndarray
+    _B: np.ndarray
+    _AB: np.ndarray
+    _f_A: np.ndarray
+    _f_B: np.ndarray
+    _f_AB: np.ndarray
```

Comment:
Actually quite interesting to save. In other libraries, it's a common ask that people want to reuse samples they previously generated to do something else. Like another analysis, constructing a model, KDE to do uncertainty propagation, do some visualisation (like I did in the example, or a cobweb, these 2 are the most popular options), etc.

Note: at one point we might be tempted to split the main function to allow passing samples instead of letting it create and make the indices. I would refrain to do that because it's extremely common for people to make a mistake when generating the matrices. You have no means of checking it besides the shape. e.g. in SALib they have exactly this, and I have seen bug reports of weird values, and the root cause was just a wrong sample being passed to the function.

What could be more interesting as a follow up, would be to have ways to increase the sample size to do a convergence analysis.